File: tools/slicec-cs/src/generators/struct_generator.rs
// Copyright (c) ZeroC, Inc.

use crate::builders::{AttributeBuilder, Builder, CommentBuilder, ContainerBuilder, FunctionBuilder, FunctionType};
use crate::code_block::CodeBlock;
use crate::cs_attributes::CsReadonly;
use crate::decoding::*;
use crate::encoding::*;
use crate::member_util::*;
use crate::slicec_ext::{CommentExt, EntityExt, MemberExt, TypeRefExt};
use slicec::grammar::*;
use slicec::supported_encodings::SupportedEncodings;

pub fn generate_struct(struct_def: &Struct) -> CodeBlock {
    let escaped_identifier = struct_def.escape_identifier();
    let fields = struct_def.fields();
    let namespace = struct_def.namespace();

    let mut declaration = vec![struct_def.access_modifier()];
    if struct_def.has_attribute::<CsReadonly>() {
        declaration.push("readonly");
    }
    declaration.extend(["partial", "record", "struct"]);

    let mut builder = ContainerBuilder::new(&declaration.join(" "), &escaped_identifier);
    if let Some(summary) = struct_def.formatted_doc_comment_summary() {
        builder.add_comment("summary", summary);
    }
    builder
        .add_generated_remark("record struct", struct_def)
        .add_comments(struct_def.formatted_doc_comment_seealso())
        .add_obsolete_attribute(struct_def);

    builder.add_block(
        fields
            .iter()
            .map(|m| field_declaration(m))
            .collect::<Vec<_>>()
            .join("\n\n")
            .into(),
    );

    let mut main_constructor = FunctionBuilder::new(
        struct_def.access_modifier(),
        "",
        &escaped_identifier,
        FunctionType::BlockBody,
    );
    main_constructor.add_comment(
        "summary",
        format!(r#"Constructs a new instance of <see cref="{escaped_identifier}" />."#),
    );

    for field in &fields {
        main_constructor.add_parameter(
            &field.data_type().field_type_string(&namespace),
            field.parameter_name().as_str(),
            None,
            field.formatted_doc_comment_summary(),
        );
    }
    main_constructor.set_body({
        let mut code = CodeBlock::default();
        for field in &fields {
            writeln!(code, "this.{} = {};", field.field_name(), field.parameter_name(),);
        }
        code
    });
    builder.add_block(main_constructor.build());

    // Decode constructor
    let mut decode_body = generate_encoding_blocks(&fields, struct_def.supported_encodings(), decode_fields, "decoder");

    if !struct_def.is_compact {
        writeln!(decode_body, "decoder.SkipTagged();");
    }
    builder.add_block(
            FunctionBuilder::new(
                struct_def.access_modifier(),
                "",
                &escaped_identifier,
                FunctionType::BlockBody,
            )
            .add_comment(
                "summary",
                format!(r#"Constructs a new instance of <see cref="{escaped_identifier}" /> and decodes its fields from a Slice decoder."#),
            )
            .add_parameter(
                "ref SliceDecoder",
                "decoder",
                None,
                Some("The Slice decoder.".to_owned()),
            )
            .set_body(decode_body)
            .build(),
        );

    // Encode method
    let mut encode_body = generate_encoding_blocks(&fields, struct_def.supported_encodings(), encode_fields, "encoder");

    if !struct_def.is_compact {
        writeln!(encode_body, "encoder.EncodeVarInt32(Slice2Definitions.TagEndMarker);");
    }
    builder.add_block(
        FunctionBuilder::new(
            &(struct_def.access_modifier().to_owned() + " readonly"),
            "void",
            "Encode",
            FunctionType::BlockBody,
        )
        .add_comment("summary", "Encodes the fields of this struct with a Slice encoder.")
        .add_parameter(
            "ref SliceEncoder",
            "encoder",
            None,
            Some("The Slice encoder.".to_owned()),
        )
        .set_body(encode_body)
        .build(),
    );

    builder.build()
}

/// Generates an expression for encoding or decoding the fields of a struct.
/// It checks which encodings this struct supports, and only generates code for the necessary encodings.
fn generate_encoding_blocks(
    fields: &[&Field],
    encodings: SupportedEncodings,
    encoding_fn: fn(&[&Field], Encoding) -> CodeBlock,
    encoding_source: &'static str,
) -> CodeBlock {
    match encodings[..] {
        [] => unreachable!("No supported encodings"),
        [encoding] => encoding_fn(fields, encoding),
        _ => {
            let slice1_block = encoding_fn(fields, Encoding::Slice1);
            let slice2_block = encoding_fn(fields, Encoding::Slice2);

            // The encoding blocks are only empty for empty structs. But `Slice1` doesn't support empty structs.
            // So this branch of the match statement is never hit, since it's for structs that support both encodings.
            assert!(!slice1_block.is_empty() && !slice2_block.is_empty());

            // Only write one encoding block if `slice1_block` and `slice2_block` are the same.
            if slice1_block.to_string() == slice2_block.to_string() {
                return slice2_block;
            }

            format!(
                "\
if ({encoding_source}.Encoding == SliceEncoding.Slice1)
{{
{slice1_block}
}}
else // Slice2
{{
{slice2_block}
}}
",
                slice1_block = slice1_block.indent(),
                slice2_block = slice2_block.indent(),
            )
            .into()
        }
    }
}

File: src/ZeroC.Slice.Generator/StructGenerator.cs
// Copyright (c) ZeroC, Inc.

using ZeroC.CodeBuilder;
using ZeroC.Slice.Symbols;

namespace ZeroC.Slice.Generator;

/// <summary>Generates C# record structs from Slice struct definitions.</summary>
internal static class StructGenerator
{
    internal static CodeBlock Generate(Struct structDef)
    {
        string identifier = structDef.Name;
        string currentNamespace = structDef.Namespace;
        string accessModifier = structDef.AccessModifier;
        bool isReadonly = structDef.Attributes.HasAttribute(CSAttributes.CSReadonly);

        // Build the declaration prefix.
        string declaration = isReadonly
            ? $"{accessModifier} readonly partial record struct"
            : $"{accessModifier} partial record struct";

        string scopedId = structDef.ScopedIdentifier;

        return new ContainerBuilder(declaration, identifier)
            .AddDocCommentSummary(structDef.Comment, currentNamespace)
            .AddComment(
                "remarks",
                $"The Slice compiler generated this record struct from the Slice struct <c>{scopedId}</c>.")
            .AddDocCommentSeeAlso(structDef.Comment, currentNamespace)
            .AddBlock(CodeBlock.FromBlocks(
                structDef.Fields.Select(
                    f => FieldDeclaration(f, currentNamespace, accessModifier, isReadonly))))
            .AddBlock(GenerateMainConstructor(structDef, currentNamespace, identifier, accessModifier))
            .AddBlock(GenerateDecodeConstructor(structDef, currentNamespace, identifier, accessModifier))
            .AddBlock(GenerateEncodeMethod(structDef, currentNamespace, accessModifier))
            .Build();
    }

    private static CodeBlock GenerateMainConstructor(
        Struct structDef,
        string currentNamespace,
        string identifier,
        string accessModifier)
    {
        bool hasRequiredField = structDef.Fields.Any(f => f.IsRequired);

        FunctionBuilder ctor = new FunctionBuilder(accessModifier, "", identifier, FunctionType.BlockBody)
            .AddComment("summary", @$"Constructs a new instance of <see cref=""{identifier}"" />.");

        if (hasRequiredField)
        {
            ctor.AddSetsRequiredMembersAttribute();
        }

        foreach (Field field in structDef.Fields)
        {
            string typeString = field.DataType.FieldTypeString(field.DataTypeIsOptional, currentNamespace);
            string paramName = field.ParameterName;
            ctor.AddParameter(typeString, paramName);
        }

        var body = new CodeBlock();
        foreach (Field field in structDef.Fields)
        {
            body.WriteLine($"this.{field.Name} = {field.ParameterName};");
        }
        ctor.SetBody(body);

        return ctor.Build();
    }

    private static CodeBlock GenerateDecodeConstructor(
        Struct structDef,
        string currentNamespace,
        string identifier,
        string accessModifier)
    {
        IReadOnlyList<Field> sortedFields = structDef.Fields.GetSortedFields();
        bool hasRequiredField = structDef.Fields.Any(f => f.IsRequired);

        FunctionBuilder ctor = new FunctionBuilder(accessModifier, "", identifier, FunctionType.BlockBody)
            .AddComment(
                "summary",
                @$"Constructs a new instance of <see cref=""{identifier}"" /> and decodes its fields from a Slice decoder.")
            .AddComment("param", "name", "decoder", "The Slice decoder.")
            .AddParameter("ref SliceDecoder", "decoder");

        if (hasRequiredField)
        {
            ctor.AddSetsRequiredMembersAttribute();
        }

        var body = new CodeBlock();

        int bitSequenceSize = structDef.Fields.GetBitSequenceSize();
        if (bitSequenceSize > 0)
        {
            body.WriteLine($"var bitSequenceReader = decoder.GetBitSequenceReader({bitSequenceSize});");
        }

        foreach (Field field in sortedFields)
        {
            string fieldName = field.Name;
            string decodeExpr = field.GetFieldDecodeExpression(currentNamespace);
            body.WriteLine($"this.{fieldName} = {decodeExpr};");
        }

        if (!structDef.IsCompact)
        {
            body.WriteLine("decoder.SkipTagged();");
        }

        ctor.SetBody(body);
        return ctor.Build();
    }

    private static CodeBlock GenerateEncodeMethod(Struct structDef, string currentNamespace, string accessModifier) =>
        new FunctionBuilder($"{accessModifier} readonly", "void", "Encode", FunctionType.BlockBody)
            .AddComment("summary", "Encodes the fields of this struct with a Slice encoder.")
            .AddComment("param", "name", "encoder", "The Slice encoder.")
            .AddParameter("ref SliceEncoder", "encoder")
            .SetBody(structDef.Fields.GenerateEncodeBody(
                currentNamespace,
                includeTagEndMarker: !structDef.IsCompact))
            .Build();


    private static CodeBlock FieldDeclaration(
        Field field,
        string currentNamespace,
        string accessModifier,
        bool parentReadonly)
    {
        var code = new CodeBlock();

        code.WriteDocCommentSummary(field.Comment, currentNamespace);

        // cs::attribute
        code.WriteCSAttributes(field.Attributes);

        string typeString = field.DataType.FieldTypeString(field.DataTypeIsOptional, currentNamespace);
        string fieldName = field.Name;
        string required = field.IsRequired ? "required " : "";
        bool fieldReadonly = field.IsReadonly;
        string accessor = (parentReadonly || fieldReadonly) ? "{ get; init; }" : "{ get; set; }";

        code.WriteLine($"{accessModifier} {required}{typeString} {fieldName} {accessor}");

        return code;
    }
}
